PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,508 @@
+            } else if finishable {
+                maybeCompletion?(transaction, nil, maybeError, false)
+                if self.finishTransactions {
+                    self.storeKitWrapper.finishTransaction(transaction)
+                }
+            } else {
+                Logger.error(Strings.receipt.unknown_backend_error)
+                maybeCompletion?(transaction, nil, maybeError, false)
+            }
+        }
+    }
+
+    func markSyncedIfNeeded(subscriberAttributes: SubscriberAttributeDict?, appUserID: String, maybeError: Error?) {
+        if let error = maybeError as NSError? {
+            if !error.rc_successfullySynced {
+                return
+            }
+            let attributeErrors = (error.rc_subscriberAttributesErrors?.debugDescription ?? "None") as NSString
+            Logger.error(String(format: Strings.attribution.subscriber_attributes_error, attributeErrors))
+        }
+
+        subscriberAttributesManager.markAttributesAsSynced(subscriberAttributes, appUserID: appUserID)
+    }
+
+    func syncPurchases(receiptRefreshPolicy: ReceiptRefreshPolicy,
+                       isRestore: Bool,
+                       maybeCompletion: ((PurchaserInfo?, Error?) -> Void)?) {
+        if !self.allowSharingAppStoreAccount {
+            Logger.warn(Strings.restore.restoretransactions_called_with_allow_sharing_appstore_account_false_warning)
+        }
+
+        let currentAppUserID = appUserID
+        let unsyncedAttributes = unsyncedAttributes
+        // Refresh the receipt and post to backend, this will allow the transactions to be transferred.
+        // https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/StoreKitGuide/Chapters/Restoring.html
+        receiptFetcher.receiptData(refreshPolicy: receiptRefreshPolicy) { maybeReceiptData in
+            guard let receiptData = maybeReceiptData,
+                  !receiptData.isEmpty else {
+                      if SystemInfo.isSandbox {
+                          Logger.appleWarning(Strings.receipt.no_sandbox_receipt_restore)
+                      }
+
+                      if let completion = maybeCompletion {
+                          completion(nil, ErrorUtils.missingReceiptFileError())
+                      }
+                      return
+                  }
+
+            let maybeCachedPurchaserInfo = self.purchaserInfoManager.cachedPurchaserInfo(appUserID: currentAppUserID)
+            let hasOriginalPurchaseDate = maybeCachedPurchaserInfo?.originalPurchaseDate != nil
+            let receiptHasTransactions = self.receiptParser.receiptHasTransactions(receiptData: receiptData)
+
+            if !receiptHasTransactions && hasOriginalPurchaseDate {
+                if let completion = maybeCompletion {
+                    self.operationDispatcher.dispatchOnMainThread {
+                        completion(maybeCachedPurchaserInfo, nil)
+                    }
+                }
+                return
```

Comment:
nitpick: this would perhaps be nicer as a guard-let since it checks a condition and early returns if we're not ready to move forward

Reply:
I think it reads better this way. Here's what the guard looks like:

```
guard receiptHasTransactions || !hasOriginalPurchaseDate else {
```

IMO guards are best when you have `&&` clauses, when you start throwing in `||` as well as `!` it becomes kinda difficult to reason about.